### PR TITLE
Update v4 content to run from v4 data

### PR DESCRIPTION
**Pull request by georgf**
This updates the remote content here to be powered by the v4 client data (see [bug 1150491](https://bugzilla.mozilla.org/show_bug.cgi?id=1150491) for more background).

Simple testing steps:
* checkout
* build from top source dir: `./generate.py --output-dir html -f`
* start server in top source dir: `python -m SimpleHTTPServer 8000`
* go to `about:config` and change `datareporting.healthreport.about.reportUrl` to `http://localhost:8000/html/%LOCALE%/v4/`
* visit `about:healthreport`

Note that the v4 unit test currently is not integrated in the grunt build due to complications (old PhantomJS instance vs newer JS & no proper error output).
You can run it manually for now by navigating to `http://localhost:8000/tests/unit_v4.html`.
We are ok with running it manually for now to avoid blocking this bug, medium-term we want to switch this over to more modern tests (probably Node.js based) that have better support within Mozilla.

### CODE REVIEW

**Comment by georgf** [comment]
@lauraxt, @lonnen, can either of you review?

**Comment by lonnen** [comment]
2 tests of 5 passed, 3 failed.

```
Expected 1 main crash this month and got 1
Expected 2 main crashes in the last 30 days and got 1
Expected: 	

2

Result: 	

1

Diff: 	

2 1 

Source: 	

@http://localhost:8000/tests/lib/unit_v4.js:28:9
jQuery.Callbacks/fire@http://code.jquery.com/jquery-1.9.0.js:1017:10
jQuery.Callbacks/self.fireWith@http://code.jquery.com/jquery-1.9.0.js:1127:7
done@http://code.jquery.com/jquery-1.9.0.js:8021:5
.send/callback@http://code.jquery.com/jquery-1.9.0.js:8558:8
EventHandlerNonNull*.send@http://code.jquery.com/jquery-1.9.0.js:8581:7
.ajax@http://code.jquery.com/jquery-1.9.0.js:7930:5
jQuery[method]@http://code.jquery.com/jquery-1.9.0.js:7569:1
.getJSON@http://code.jquery.com/jquery-1.9.0.js:8055:10
@http://localhost:8000/tests/lib/unit_v4.js:20:5

Expected 0 plugin crashes this month and got 0
Expected 2 sessions this month and got 1
Expected: 	

2

Result: 	

1

Diff: 	

2 1 

Source: 	

@http://localhost:8000/tests/lib/unit_v4.js:31:9
jQuery.Callbacks/fire@http://code.jquery.com/jquery-1.9.0.js:1017:10
jQuery.Callbacks/self.fireWith@http://code.jquery.com/jquery-1.9.0.js:1127:7
done@http://code.jquery.com/jquery-1.9.0.js:8021:5
.send/callback@http://code.jquery.com/jquery-1.9.0.js:8558:8
EventHandlerNonNull*.send@http://code.jquery.com/jquery-1.9.0.js:8581:7
.ajax@http://code.jquery.com/jquery-1.9.0.js:7930:5
jQuery[method]@http://code.jquery.com/jquery-1.9.0.js:7569:1
.getJSON@http://code.jquery.com/jquery-1.9.0.js:8055:10
@http://localhost:8000/tests/lib/unit_v4.js:20:5

Expected 10 hours session time this month and got 5
Expected: 	

36000

Result: 	

18000

Diff: 	

36000 18000 

Source: 	

@http://localhost:8000/tests/lib/unit_v4.js:32:9
jQuery.Callbacks/fire@http://code.jquery.com/jquery-1.9.0.js:1017:10
jQuery.Callbacks/self.fireWith@http://code.jquery.com/jquery-1.9.0.js:1127:7
done@http://code.jquery.com/jquery-1.9.0.js:8021:5
.send/callback@http://code.jquery.com/jquery-1.9.0.js:8558:8
EventHandlerNonNull*.send@http://code.jquery.com/jquery-1.9.0.js:8581:7
.ajax@http://code.jquery.com/jquery-1.9.0.js:7930:5
jQuery[method]@http://code.jquery.com/jquery-1.9.0.js:7569:1
.getJSON@http://code.jquery.com/jquery-1.9.0.js:8055:10
@http://localhost:8000/tests/lib/unit_v4.js:20:5

```

File: js/data_v4.js [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+
+    if (currentYear === year && currentMonth === month) {
+        return true;
+    }
+    return false;
+};
+
+function isPastNDays(day, n, now) {
+    if (!now) {
+        now = new Date();
```
Comment (lonnen) [inline js/data_v4.js]: `now` appears to be unused in this function

File: js/data_v4.js [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+// Is this the first load for the document?
+var isFirstLoad = true;
+
+// Converts the date passed to a Date object and checks
+// whether the current month is equal to the month of the
+// day argument passed.
+var isCurrentMonth = function(day, now) {
+    if (!now) {
+        now = new Date();
```
Comment (lonnen) [inline js/data_v4.js]: `now` appears to be unused in this function

File: js/data_v4.js [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    var difference = new Date().getTime() - new Date(day).getTime();
+    return difference < ONE_DAY * n;
+}
+
+function populateEnvironment(environment) {
+    var vitalStatsValueContainers = $('#vital_stats .statsBoxSection-value');
+
+    var channel = environment.settings.update.channel;
+    var updates = environment.settings.update.enabled ?
+      (environment.settings.update.autoDownload ? "automatic" : "prompt") : "disabled";
```
Comment (lonnen) [inline js/data_v4.js]: nested multi-line ternaries make this unnecessarily difficult to read. recommend rewriting these as expanded if/else

**Comment by lonnen** [comment]
A more thorough review is going to move slower. You're using a bunch of modern JS that I'm not well experienced with, and promises have never been an intuitive abstraction for me. You'd probably be better off finding a different reviewer for the rest of it.

If you and kparlante are reviewing each other on this, that's alright. As far as I'm concerned you two own this at this point -- you're certainly far more familiar with it than I ever was.

**Comment by lonnen** [comment]
```
Expected 1 main crash this month and got 1
Expected 2 main crashes in the last 30 days and got 2
Expected 0 plugin crashes this month and got 0
Expected 2 sessions this month and got 1

Expected: 2
Result: 1

Diff: 	2 1 

Source: 	
@http://localhost:8000/tests/lib/unit_v4.js:31:9
jQuery.Callbacks/fire@http://code.jquery.com/jquery-1.9.0.js:1017:10
jQuery.Callbacks/self.fireWith@http://code.jquery.com/jquery-1.9.0.js:1127:7
done@http://code.jquery.com/jquery-1.9.0.js:8021:5
.send/callback@http://code.jquery.com/jquery-1.9.0.js:8558:8
EventHandlerNonNull*.send@http://code.jquery.com/jquery-1.9.0.js:8581:7
.ajax@http://code.jquery.com/jquery-1.9.0.js:7930:5
jQuery[method]@http://code.jquery.com/jquery-1.9.0.js:7569:1
.getJSON@http://code.jquery.com/jquery-1.9.0.js:8055:10
@http://localhost:8000/tests/lib/unit_v4.js:20:5

Expected 10 hours session time this month and got 5
Expected: 36000
Result: 18000

Diff: 	36000 18000 

Source: 	
@http://localhost:8000/tests/lib/unit_v4.js:32:9
jQuery.Callbacks/fire@http://code.jquery.com/jquery-1.9.0.js:1017:10
jQuery.Callbacks/self.fireWith@http://code.jquery.com/jquery-1.9.0.js:1127:7
done@http://code.jquery.com/jquery-1.9.0.js:8021:5
.send/callback@http://code.jquery.com/jquery-1.9.0.js:8558:8
EventHandlerNonNull*.send@http://code.jquery.com/jquery-1.9.0.js:8581:7
.ajax@http://code.jquery.com/jquery-1.9.0.js:7930:5
jQuery[method]@http://code.jquery.com/jquery-1.9.0.js:7569:1
.getJSON@http://code.jquery.com/jquery-1.9.0.js:8055:10
@http://localhost:8000/tests/lib/unit_v4.js:20:5
```

Two failures with the new revisions. You fixed the 30 days measurements. Session counts are still wrong.

**Comment by georgf** [comment]
Strange, it runs fine here, i wonder if that's locale/timezone specific, will investigate tomorrow - thanks lonnen.

Moving forward, i will drive the update here and Katie can do in-depth reviews. Hope that works fine for everyone.

File: js/fhr_v4.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        let min = Math.min.apply(null, [for (time of startupTimes) time[0]]);
+        let max = Math.max.apply(null, [for (time of startupTimes) time[0]]);
+        let minoffset = min % ONE_DAY;
+        let ticks = [];
+        let interval = ONE_DAY;
+        if ((max - min) / interval > 20) {
+            interval = ONE_DAY * 2;
+        }
+        for (let tick = min - minoffset + new Date().getTimezoneOffset() * 60 * 1000;
+             tick < max; tick += interval) {
```
Comment (kparlante) [inline js/fhr_v4.js]: No tick for the most recent measures. tick <= max looks better with my data

**Comment by kparlante** [comment]
Tests pass; changes made by @georgf look reasonable.
A nit -- the graph doesn't have a tick for the most recent date.
screen-1
vs
screen-2

**Comment by georgf** [comment]
Those last revisions should make the startup times table much robust.

The one UX nit i see is that startup times collected in different timezones than the current one don't line up on the days exactly... We might want to wrap data points to the closest days per the current timezone.
But that seems like a minor or negligible issue we could think about later.

**Comment by kparlante** [comment]
> But that seems like a minor or negligible issue we could think about later.

Agreed. Another minor nit: the "raw data" view should probably have a margin on the right hand side. I wouldn't block on it.

screen-3

**Comment by kparlante** [comment]
r+ tests pass, table of startup times looks good. Lets get this to staging...

**Comment by lonnen** [comment]
After you merge this grab the sha and webops can help you push it. Details here: https://mana.mozilla.org/wiki/pages/viewpage.action?pageId=33101035#fhr.cdn.mozilla.net%28FirefoxHealthReport%29-DeploymentBugs

**Comment by georgf** [comment]
Who can merge this PR? @kparlante or @lonnen?

**Pull request merged**